### PR TITLE
Fix ampersand display issue

### PR DESCRIPTION
**Pull request by mickohanlon23**
"**Janssen (J&J) COVID-19 Vaccine**" was displaying as "**Janssen (J\&amp;J) COVID-19 Vaccine**"

### CODE REVIEW

File: src/components/HealthCardDisplay/HealthCardDisplay.js [hunk 1]
```diff
@@ -74,9 +74,9 @@ const HealthCardDisplay = () => {
       if (!cvxCodes[coding.code]) {
         return coding.system ? `${coding.system}#${coding.code}` : coding.code;
       }
-      return cvxCodes[coding.code];
+      return cvxCodes[coding.code].replace(/&amp;/g, '&');
```
Comment (nkarip) [inline src/components/HealthCardDisplay/HealthCardDisplay.js:77]: Would move this into the iisstandards.js file where the XML handling happens. Also would consider using a replacer function like https://stackoverflow.com/questions/7918868/how-to-escape-xml-entities-in-javascript to handle all the cases (not just &).

File: src/components/HealthCardDisplay/HealthCardDisplay.js [hunk 2]
```diff
@@ -74,9 +74,9 @@ const HealthCardDisplay = () => {
...
     }
-    return tradenames[coding.code];
+    return tradenames[coding.code].replace(/&amp;/g, '&');
```
Comment (nkarip) [inline src/components/HealthCardDisplay/HealthCardDisplay.js:79]: Same as above.

**Review by nkarip — CHANGES_REQUESTED** [review]

File: src/components/HealthCardDisplay/HealthCardDisplay.js (hunk 2 above)
Comment (mickohanlon23) [reply src/components/HealthCardDisplay/HealthCardDisplay.js]: Moved HTML entity decoding to `iisstandards.js` (same as above).

**Review by radamson — CHANGES_REQUESTED** [review]
I tested on main and it looks like the real issue is that these characters are unneccesarily escaped in the XML before being parsed. Instead of escaping and then unescaping, just don't escape in the first place. As an added bonus this appears to fix the Firefox issue!

**TL;dR:**

The only change you need is to delete this:

https://github.com/dvci/shc-web-verifier/blob/385c8641c4a6024c0b646e0e25d4f43e2165ed3c/src/components/HealthCardDisplay/iisstandards.js#L13-L17

<details>
<summary>Test QR Code and results after deleting the character escape codes</summary>

jj

Screen Shot 2022-02-02 at 12 48 40 PM

</details>

**Comment by mickohanlon23** [comment]
> I tested on main and it looks like the real issue is that these characters are unneccesarily escaped in the XML before being parsed. Instead of escaping and then unescaping, just don't escape in the first place. As an added bonus this appears to fix the Firefox issue!
> 
> **TL;DR:**
> 
> Just delete this. No other changes required:
> 
> https://github.com/dvci/shc-web-verifier/blob/385c8641c4a6024c0b646e0e25d4f43e2165ed3c/src/components/HealthCardDisplay/iisstandards.js#L13-L17
> 
> Test QR Code and results after deleting the character escape codes

I had proposed that same fix in a discussion with Neelima offline, but she mentioned that there were some issues with the way the XML files are formatted that led to errors without the manual encoding. @nkarip, do you want to weigh in?

File: src/components/HealthCardDisplay/iisstandards.js [hunk 3]
```diff
@@ -5,6 +5,15 @@ import cvxXml from './iisstandards_cvx.xml';
 let tradenames = null;
 let cvx = null;
 
+function decodeEntity(xmlString) {
```
Comment (nkarip) [inline src/components/HealthCardDisplay/iisstandards.js:8]: This looks good. Would also create an encodeEntity that is the same in reverse, to keep it clean.

**Review by nkarip — CHANGES_REQUESTED** [review]

File: src/components/HealthCardDisplay/iisstandards.js (hunk 3 above)
Comment (nkarip) [reply src/components/HealthCardDisplay/iisstandards.js:8]: Also would add tests to confirm this functionality, see iisstandards.test.js in main branch for jest examples.

**Comment by nkarip** [comment]
@radamson the CDC files use reserved characters without either a CDATA block or escaping. This was causing DOMParser errors.

**Comment by radamson** [comment]
@nkarip this method is only used with the local `iisstandards_cvx.xml` and `iisstandards_tradenames.xml` files in the repo. The ampersands appear to be properly escaped in all cases in these files:

https://github.com/dvci/shc-web-verifier/blob/d3f64086fa8277f96e180d41c53e580638e0a50c/src/components/HealthCardDisplay/iisstandards_cvx.xml#L145

https://github.com/dvci/shc-web-verifier/blob/d3f64086fa8277f96e180d41c53e580638e0a50c/src/components/HealthCardDisplay/iisstandards_tradename.xml#L1553

The `replace` invocations are causing these to be double escaped which cause issues with parsing and are the reason for ampersands showing up improperly.

Can you point to me to an example of these characters not being escaped properly or how to produce a DOM error with removing this code?

**Comment by nkarip** [comment]
See previous version of file at https://github.com/dvci/shc-web-verifier/blob/eefdbf6658264b870818d936184966e7a5e8a2a5/src/components/HealthCardDisplay/iisstandards_tradename.xml (J&J is not escaped). While they fixed it, there is not guarantee they won't re-introduce errors at some point.

**Comment by radamson** [comment]
I actually don't see the Firefox issue on the current main branch, but if I swap in the invalid unescaped XML you linked I see:

Screen Shot 2022-02-14 at 11 33 41 AM

I think this error is coming from `axios.get(file...` so we'll still have this Firefox error as long as we're committing invalid XML.

My preference would be to only commit valid XML files and do no custom XML handling/escaping.

If we really don't want to do this and want to allow unescaped ampersands in the CDC XML files, then we could still achieve the same fix with a single line of code by making the original `replace` regex more precise.

i.e. change

https://github.com/dvci/shc-web-verifier/blob/d3f64086fa8277f96e180d41c53e580638e0a50c/src/components/HealthCardDisplay/iisstandards.js#L15

to:
```javascript
.replace(/&(?!amp;)/g, '&amp;')
````

That way correctly escaped ampersands aren't re-escaped (which is causing the current issue since (`&amp;` turns into `&amp;amp` which then gets shown as `&amp;` in the UI).

**Comment by nkarip** [comment]
Ah, that makes sense. I would favor fixing the regex here. We do currently have some tests in [iisstandards.test.js](https://github.com/dvci/shc-web-verifier/blob/main/src/components/HealthCardDisplay/iisstandards.test.js) that should fail if the latest XML files have DOMParser errors but since we're updating the files in real time I would prefer to have some mitigation to handle potential known issues. Other option is to change the github workflow so it doesn't update the main files but creates a branch instead, so we don't inadvertently introduce bad XML.

**Comment by radamson** [comment]
Looking through recent commits it doesn't look like the CDC XML has changed frequently or had this issue consistently so making the github workflow create a PR instead would be the best approach IMO especially since there could be other unforeseen issues. That way we'd have less code to maintain and special cases to consider as well.

**Comment by nkarip** [comment]
I guess the only problem with creating a PR is that someone will have to monitor the repo long-term for these PRs, and review and merge quickly. I don't get notifications automatically when either update workflow runs fail or new PRs are entered, so we'd have to set this up as well.

**Comment by radamson** [comment]
Yup, its a little more work to do that so we can save it for a separate PR and get this fix in now. I'll create a follow on ticket to update the github action so we can prevent future console errors in Firefox or catch other unexpected changes (e.g. URL changes since it looks like they're using a different sub-domain for this page for load balancing)

@mickohanlon23 
**This PR:**
Update the initial regexes to be more specific:

i.e. change

https://github.com/dvci/shc-web-verifier/blob/d3f64086fa8277f96e180d41c53e580638e0a50c/src/components/HealthCardDisplay/iisstandards.js#L15

to:
```javascript
.replace(/&(?!amp;)/g, '&amp;')
````

**Comment by mickohanlon23** [comment]
> That way correctly escaped ampersands aren't matched and escaped twice (which is causing the current issue since (`&amp;` turns into `&amp;amp` which then gets shown as `&amp;` in the UI).

**Comment by mickohanlon23** [comment]
@radamson @nkarip 

> Yup, its a little more work to do that so we can save it for a separate PR and get this fix in now. I'll create a follow on ticket to update the github action so we can prevent future console errors in Firefox or catch other unexpected changes (e.g. URL changes since it looks like they're using a different sub-domain for this page for load balancing)
> 
> @mickohanlon23 **This PR:** Update the initial regexes to be more specific:
> 
> i.e. change
> 
> https://github.com/dvci/shc-web-verifier/blob/d3f64086fa8277f96e180d41c53e580638e0a50c/src/components/HealthCardDisplay/iisstandards.js#L15
> 
> to:
> 
> ```js
> .replace(/&(?!amp;)/g, '&amp;')
> ```

@radamson @nkarip Makes sense. I specified the regex that looks for '&', as suggested. Can somebody confirm that this does not cause errors when run using Firefox?

File: src/components/HealthCardDisplay/iisstandards.js [hunk 4]
```diff
@@ -12,7 +12,7 @@ async function fetchCdcXml(file) {
   let data = await response.data;
   data = data
     .replace(/<\?xml.*\?>/g, '')
-    .replace(/&/g, '&amp;')
+    .replace(/&(?!amp;)/g, '&amp;')
     .replace(/"/g, '&quot;')
     .replace(/'/g, '&apos;');
```
Comment (radamson) [inline src/components/HealthCardDisplay/iisstandards.js:17]: Let's do them all while we're at it
```suggestion
    .replace(/"(?!quot;)/g, '&quot;')
    .replace(/'(?!apos;)/g, '&apos;');
```

Comment (mickohanlon23) [reply src/components/HealthCardDisplay/iisstandards.js:17]: Wouldn't this not be relevant here, since the entity in this instance wouldn't actually contain a quote or an apostrophe character?

Comment (radamson) [reply src/components/HealthCardDisplay/iisstandards.js:17]: Ah, yup you're right.

I'm still a little wary about `.replace(/"/g, '&quot;')` since it could break XML properties outside the name, but we can handle that when preventing bad XML from being entered.

**Review by radamson — APPROVED** [review]

File: src/components/HealthCardDisplay/iisstandards.js (hunk 3 above)
Comment (mickohanlon23) [reply src/components/HealthCardDisplay/iisstandards.js]: The `decodeEntity` function no longer exists, per discussion below. I did move the `.replace`s used for the initial character escaping into a separate function so that they could be tested in isolation against mock XML data.

Comment (nkarip) [reply src/components/HealthCardDisplay/iisstandards.js]: Updated with a test to check the reverse, that we're not encoded already encoded characters. This failed, which required a fix to the regex.

**Review by nkarip — APPROVED** [review]
